tmf/org.eclipse.tracecompass.tmf.core/src/org/eclipse/tracecompass/tmf/core/dataprovider/DataProviderManager.java

```diff
@@ -272,4 +272,53 @@ public synchronized Collection<IDataProviderFactory> getFactories() {
         String factoryId = ids.length > 1 ? ids[0] : id;
         return fDataProviderFactories.get(factoryId);
     }
+
+    /**
+     * Add a new data provider factory.
+     *
+     * If a data provider factory is associated with the ID already exists
+     * it will replace and return the existing data provider factory.
+     *
+     * @param id
+     *            The data provider factory ID
+     * @param factory
+     *            The data provider factory implementation
+     * @return the previous data provider factory associated with the key or null
+     *
+     * @since 9.4
+     */
+    public synchronized @Nullable IDataProviderFactory addDataProviderFactory(String id, IDataProviderFactory factory) {
+        return fDataProviderFactories.put(id, factory);
+    }
+
+    /**
+     * Remove a data provider from the instances by ID. This method will not
+     * dispose of the data provider. It is the responsibility of the caller to
+     * dispose of it if needed.
+     *
+     * @param trace
+     *            The trace for which to remove the data provider
+     * @param id
+     *            The The ID of the data provider to remove
+     * @return {@code true} if any elements were removed
+     * @since 9.4
+     */
+    public boolean removeDataProvider(ITmfTrace trace, String id) {
```

Comment:
Would it be reasonable to consider that when we replace or remove a data provider factory, then all instances that were created with that id would be automatically removed from here (for every trace)?

Also I wonder since we have no control over who has a reference to those data providers, how could any entity be sure that it would be ok to dispose of it? Probably just the owner of the data provider, who would be unaware that its factory was ever removed. It can continue to use the data provider until it no longer needs it?

Reply:
> Would it be reasonable to consider that when we replace or remove a data provider factory, then all instances that were created with that id would be automatically removed from here (for every trace)?
> 
I implemented it the same way as the other removeDataProvider() method (line 250) which is used with ScriptedDataProviderManager. The new method is used for similar use case: Some entity is creating data provider which is registered to the DataProviderManager and the same entity is responsible to remove it. 

I get your point. Any suggestion how to implement that differently for this use case?

BTW, maybe we should return the original data provider if it exists else null, to be consistent to the other data provider method. WDYT?

> Also I wonder since we have no control over who has a reference to those data providers, how could any entity be sure that it would be ok to dispose of it? Probably just the owner of the data provider, who would be unaware that its factory was ever removed. It can continue to use the data provider until it no longer needs it?

No, we don't have any control over who has a reference on the data provider. As stated a few lines above, the use case is that an entity is create a data provider and is responsible to remove and dispose it. Please let me know if you have any suggestion on how to handle that use case.

Reply:
As per our chat, I'll remvoe the removeDataProvider(trace, id) since it's not needed, and I'll update add/removeDataProviderFactory() to remove data providers instances for all traces if there are conflicts.